src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeReader.cs

```diff
@@ -259,6 +259,141 @@ public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancella
+                            return;
+                        }
+                    }
+                }
+                finally
+                {
+                    // Advance even if WriteAsync throws so the PipeReader is not left in the
+                    // currently reading state
+                    if (segment != null)
+                    {
+                        AdvanceTo(segment, segment.End, segment, segment.End);
+                    }
+                }
+            }
+
+            if (_isStreamCompleted)
+            {
+                return;
+            }
+
+            CancellationTokenRegistration reg = default;
+            if (cancellationToken.CanBeCanceled)
+            {
+                reg = cancellationToken.UnsafeRegister(state => ((StreamPipeReader)state!).Cancel(), this);
+            }
+
+            using (reg)
+            {
+                try
+                {
+                    await InnerStream.CopyToAsync(destination, tokenSource.Token).ConfigureAwait(false);
+                }
+                catch (OperationCanceledException)
+                {
+                    ClearCancellationToken();
+
+                    throw;
+                }
+            }
+        }
+
+        /// <inheritdoc />
+        public override async Task CopyToAsync(Stream destination, CancellationToken cancellationToken = default)
+        {
+            ThrowIfCompleted();
+
+            // PERF: store InternalTokenSource locally to avoid querying it twice (which acquires a lock)
+            CancellationTokenSource tokenSource = InternalTokenSource;
+            if (tokenSource.IsCancellationRequested)
+            {
+                ThrowHelper.ThrowOperationCanceledException_ReadCanceled();
+            }
+
+            if (_bufferedBytes > 0 && (!_examinedEverything || _isStreamCompleted))
+            {
+                BufferSegment? segment = _readHead;
+                try
+                {
+                    while (segment != null)
+                    {
+                        await destination.WriteAsync(segment.AvailableMemory, cancellationToken).ConfigureAwait(false);
+
+                        segment = segment.NextSegment;
+                    }
+                }
+                finally
+                {
+                    // Advance even if WriteAsync throws so the PipeReader is not left in the
+                    // currently reading state
+                    if (segment != null)
+                    {
+                        AdvanceTo(segment, segment.End, segment, segment.End);
+                    }
+                }
+            }
+
+            if (_isStreamCompleted)
+            {
+                return;
+            }
+
+            CancellationTokenRegistration reg = default;
+            if (cancellationToken.CanBeCanceled)
+            {
+                reg = cancellationToken.UnsafeRegister(state => ((StreamPipeReader)state!).Cancel(), this);
+            }
```

Comment:
I think we want to move this to the top of the method and use tokenSource .Token everywhere.

Reply:
Agreed!